experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -0,0 +1,755 @@
+      res = (res[1]*k, div(res[2], k))
+      k = k^2
+      g = gcd(k, res[2])
+   end
+   return res
+end
+
+################################################################################
+@doc Markdown.doc"""
+    cornacchia_algorithm(d::fmpz, m::fmpz)
+
+Return `true` and a solution of `x^2 + d*y^2 = m`if it exists, and false and
+`(0, 0)` otherwise.
+"""
+function cornacchia_algorithm(d::fmpz, m::fmpz)
+    (even, odd) = evenodd(m)
+    D = abs(d)
+    R = ResidueRing(ZZ, odd)
+    S = ResidueRing(ZZ, min(ZZ(8), even))
+    X = issquare_with_square_root(-R(D))
+    if !X[1]
+       return (false, (ZZ(0), ZZ(0)))
+    end
+    s1 = Hecke.data(X[2])
+    if !isone(even)
+       !isone(S(D)) && (false, (ZZ(0), ZZ(0)))
+       s2 = sqrtmod(-D, even)
+       s = abs(crt(-s1, odd, -s2+ZZ(2)^(Int(log2(even)-1)), even, true))
+    else
+       s = s1
+    end
+
+    if 2*s > m
+        B = m - s
+    else
+	    B = s
+    end
+
+    A = m
+    while B^2 >= m
+       A, B = B, A%B
+    end
+    t = m - B^2
+    L = divrem(t, D)
+    if !iszero(L[2])
+       return (false, (ZZ(0), ZZ(0)))
+    elseif !issquare(L[1])
+        return (false, (ZZ(0), ZZ(0)))
+    else
+        return (true, [B, isqrt(L[1])])
+    end
+end
+
+################################################################################
+###############Hilbert Class Polynomial#########################################
+################################################################################
+
+#Compute reduced forms of imaginary quadratic number fields
+# based on Algo 5.3.5 in the book by Cohen
+function class_data(D::fmpz)
+   D<ZZ(0) || error("Input needs to be negative")
+
+   h = 0
+   b = abs(D % 2)
+   B = root(div(-D, 3), 2)
+   L = Vector{Tuple{fmpz,fmpz,fmpz}}()
+
+   while b <= B
+      q = div(b^2-D, 4)
+      a = ZZ(max(b, 1))
+      while a^2 <= q
+         if q % a == 0 && gcd(a,b, div(q,a)) == 1
+            if a == b || a^2 == q || b == 0
+               h = h+1
+               push!(L, (a,b, div(b^2-D, 4*a)))
+            else
+               h = h + 2
+               v = div(b^2-D, 4*a)
+               push!(L, (a,-b, v))
+               push!(L, (a,b, v))
+            end
+         end
+         a = a + 1
+      end
+      b = b + 2
+   end
+   return (ZZ(h), L)
+end
+
+
+function roundarb(x::arb)
+   test = unique_integer(x)
+   test[1] && return test[2]
+   xf = floor(x)
+   x-xf < 0.5 && return xf
+   return xf + 1
+end
+
+function eta_function(tau::acb, n::Int = 17)
+   q = exppii(2*tau)
+   res = sum([(-1)^k*(q^(div(k*(3k-1), 2))+q^(div(k*(3k+1), 2))) for k in 1:n])
+   return (q, 1+res)
+end
+
+function delta_function(tau::acb, n::Int = 17)
+   (q, eta) = eta_function(tau::acb, n)
+   return q*eta^24
+end
+
+function j_function(tau::acb, n::Int = 17)
+   f_tau = inv(delta_function(tau, n))*delta_function(2*tau, n)
+   return inv(f_tau)*(256*f_tau+1)^3
+end
+
+# Computing the Hilbert class polynomial
+function hilbert_class_polynomial(D::fmpz)
```

Comment:
Why do we have/need a separate (slightly dodgy if I may say so) implementation of delta, j and class polynomials here?
Arb has had correct version of all of these for a long time.
https://arblib.org/acb_modular.html#class-polynomials

Reply:
Can I call these functions somehow in Oscar ? That's the main reason why I did it by hand, I didn't find anything... I agree that it is not too good...

Reply:
yes, just give me one second.

Reply:
for the complex functions:
```julia
julia> using Oscar

julia> CC = AcbField(100)
Complex Field with 100 bits of precision and error bounds

julia> t = sqrt(CC(-2))
[1.41421356237309504880168872421 +/- 2.41e-30]*im

julia> j_invariant(t)
[8000.00000000000000000000000 +/- 2.06e-24]

julia> modular_delta(t)
[0.0001378855139923318569712824919 +/- 3.33e-32]

julia> dedekind_eta(t)
[0.69047285586548303491090753727 +/- 3.39e-30]
```

The class polynomial is not wrapped but can be done as:

Reply:
Thank you very much ! It is the class polynomial I want.

Reply:
No problem. I was suprised this was not already wrapped in nemo. If you really wanted to do this you have to use the `arb_get_unique_fmpz` function https://github.com/fredrik-johansson/arb/blob/master/acb_modular/hilbert_class_poly.c#L81